mlflow/server/js/src/experiment-tracking/components/ExperimentView.js

```diff
@@ -254,6 +253,11 @@ export class ExperimentView extends Component {
   }
 
   setShowMultiColumns(value) {
+    ExperimentViewUtil.updateUrlWithViewState({
+      ...this.props,
+      ...this.state.persistedState,
+      showMultiColumns: value,
+    });
```

Comment:
I think since we discussed to maintain the URLState in experimentPage, updating the URLState can also be done in experimentPage as well.  So we can move `updateUrlWithViewState` function to experimentPage where we can update the URL and passed the new parse state down to experimentView. Let me know what you think about that

Reply:
@sunishsheth2009 thanks for your comments. I've now moved the `updateUrlWithViewState` function to the `ExperimentPage`. Although, I feel I'm a bit stuck on dealing with the user changing the url manually (or pressing the back button in the browser). Can you provide some pointers (or suggestions on my branch) on how to do that gracefully? 

I'll fix the failing tests after you feel my attempt is going in the right direction!

Reply:
Adding something like this in experiment page in `getDerivedStateFromProps` function might help with maintaining state.

```
if (props.location.search !== state.urlState) {
      return {
        lastRunsRefreshTime: Date.now(),
        numberOfNewRuns: 0,
        nextPageToken: null,
        persistedState: { ...getPersistedStateFromUrl(props.location.search) },
        urlState: props.location.search,
        searchRunsRequestId: getUUID(),
        ...PAGINATION_DEFAULT_STATE,
      };
    }
 ```
 
 
 also the code for 
 ```
 const getPersistedStateFromUrl = (url) => {
  const urlState = Utils.getSearchParamsFromUrl(url);

  return {
    searchInput: urlState.search === undefined ? '' : urlState.search,
    orderByKey: urlState.orderByKey === undefined ? DEFAULT_ORDER_BY_KEY : urlState.orderByKey,
    orderByAsc:
      urlState.orderByAsc === undefined ? DEFAULT_ORDER_BY_ASC : urlState.orderByAsc === 'true',
    startTime: urlState.startTime === undefined ? DEFAULT_START_TIME : urlState.startTime,
  };
};
```

This can be reused in constructor as well

Reply:
Thanks for the pointer @sunishsheth2009 ! Will give it another go.

With the goal of maintaining the state in `ExperimentPage`, wouldn't it make sense now to lift the urlState elements from `ExperimentViewPersistedState` back into `ExperimentPagePersistedState`, and pass in the needed values to `ExperimentView` as props?

Referring to:
- `showMultiColumns`
- `categorizedUncheckedKeys`
- `diffSwitchSelected`
- `preSwitchCategorizedUncheckedKeys`
- `postSwitchCategorizedUncheckedKeys`

Having the urlState spread across the two persisted state objects is quite confusing to deal with from my perspective. Or maybe I'm still missing a piece of the puzzle.

In any case, will try and take some time over the weekend to work it out

Reply:
> With the goal of maintaining the state in ExperimentPage, wouldn't it make sense now to lift the urlState elements from ExperimentViewPersistedState back into ExperimentPagePersistedState, and pass in the needed values to ExperimentView as props?

Yes that would be great! :) I like that idea of maintaining the state in one place so it can update it correctly. 
Thank you. let me know if you need any help from my side

Reply:
Thanks again for the pointer @sunishsheth2009, I've now updated the PR to maintain the urlState in `ExperimentPage`. In the end it became quite a refactor, and I've also (re-)enabled the storing of `ExperimentPage` state in localstorage. Curious to hear your what you think!